File: Practice02/Exercise01/SavingBank.cs
namespace Exercise01;

public class SavingBank
{
    public string AccountNumber { get; private set; }
    public string Cuil { get; private set; }
    public string LastName { get; private set; }
    public string FirstName { get; private set; }
    public decimal Balance { get; private set; }

    public SavingBank(string accountNumber, string cuil, string lastName, string firstName, decimal balance)
    {
        AccountNumber = accountNumber;
        Cuil = cuil;
        LastName = lastName;
        FirstName = firstName;
        Balance = balance;
    }

    public void ShowData()
    {
        Console.WriteLine("Datos de la cuenta:");
        Console.WriteLine($"\tNúmero de cuenta: {AccountNumber}");
        Console.WriteLine($"\tCuil: {Cuil}");
        Console.WriteLine($"\tApellido: {LastName}");
        Console.WriteLine($"\tNombre: {FirstName}");
        Console.WriteLine($"\tSaldo: {Balance}");
    }

    public decimal Extract(decimal amount)
    {
        if (amount <= 0)
        {
            Console.WriteLine("La cifra debe ser mayor a $0.");
            return -1;
        }

        if (amount > Balance)
        {
            Console.WriteLine("Saldo insuficiente.");
            return -1;
        }

        Balance -= amount;
        return Balance;
    }

    public decimal Deposit(decimal amount)
    {
        if (amount <= 0)
        {
            Console.WriteLine("La cifra debe ser mayor a $0.");
            return -1;
        }

        Balance += amount;
        return Balance;
    }

    public bool ValidateCuil()
    {
        if (Cuil.Length != 11)
        {
            Console.WriteLine("Cuil no válido.");
            return false;
        }

        var genre = Cuil.Substring(0, 2);
        var validationDigit = Cuil.Substring(Cuil.Length - 1, 1);
        int[] validGenres = [20, 23, 27, 30];
        int[] factorsForDniValidation = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2];

        if (!validGenres.Contains(int.Parse(genre)))
        {
            Console.WriteLine("Género no válido");
            return false;
        }

        var sum = 0;
        for (var i = 0; i < 10; i++)
        {
            sum += int.Parse(Cuil[i].ToString()) * factorsForDniValidation[i];
        }

        var module = sum % 11;
        switch (module)
        {
            case 0 when validationDigit == "0":
            case 1 when validationDigit == "9" && genre == "23":
            case 1 when validationDigit == "4" && genre == "23":
                return true;
            default:
                return int.Parse(validationDigit) == 11 - module;
        }
    }
}

File: practice-02/exercise-01/src/SavingBank.ts
export class SavingBank {
    private accountNumber: string
    private cuil: string
    private lastName: string
    private firstName: string
    private balance: number

    constructor(
        accountNumber: string,
        cuil: string,
        lastName: string,
        firstName: string,
        balance: number
    ) {
        this.accountNumber = accountNumber
        this.cuil = cuil
        this.lastName = lastName
        this.firstName = firstName
        this.balance = balance
    }
    showData(): void {}
    extract(amount: number): number | null {
        if (amount <= 0) {
            console.log('La cifra debe ser mayor a $0,00')
            return null
        }
        if (amount > this.balance) {
            console.log('Saldo insuficiente.')
            return null
        }

        this.balance -= amount
        return this.balance
    }

    deposit(amount: number): number | null {
        if (amount <= 0) {
            console.log('La cifra debe ser mayor a $0,00')
            return null
        }

        this.balance += amount
        return this.balance
    }
    validateCuil(): boolean {
        if (this.cuil.length != 11) return false

        const genre = Number(this.cuil.substring(0, 2))
        const validationDigit = Number(this.cuil.substring(10))

        if (![20, 23, 27, 30].includes(genre)) return false

        const result = this.factorsValidationResult()

        if (result == 0 && validationDigit == 0) return true

        if (result == 1 && genre == 23) {
            if (validationDigit == 9 || validationDigit == 4) return true
        }
        return 11 - result == validationDigit
    }
    get getCuil() {
        return this.cuil
    }

    private factorsValidationResult(): number {
        const validationFactors = [5, 4, 3, 2, 7, 6, 5, 4, 3, 2]
        let sum = 0

        for (let i = 0; i < validationFactors.length; i++) {
            sum += Number(this.cuil[i]) * validationFactors[i]
        }

        return sum % 11
    }
}
